packages/realm/src/app-services/SyncSession.ts

```diff
@@ -273,13 +287,16 @@ export class SyncSession {
   /** @internal */
   constructor(internal: binding.SyncSession) {
     this._internal = internal;
+    SyncSession.instances.add(this);
   }
 
   /**@internal*/
   resetInternal() {
     if (!this._internal) return;
     this._internal.$resetSharedPtr();
     this._internal = null;
+    // No need to remember this instance, now that we've already resat the internal
+    SyncSession.instances.delete(this);
```

Comment:
This makes the `resetAllInternals` loop O(n^2) because `delete()` is secretly linear. Given that you are calling `clear()` there is no need to also delete everything.

Ditto for the loop over realms in `Realm.clearTestState()` because closing a realm also calls resetInternal on its SyncSession.

PS- This is why I don't like the idea of trying to look like a well known type, while having a very different performance profile.

Reply:
It's not just the `clean()` call. `delete()` is inherently O(N) because it calls `find()`. One way around that would be to have the `IterableWeakRefs` hold something like `interface WeakFromThis<T> extends T { weakThis: WeakRef<T> }`, then delete could do `this.internal.delete(val.weakThis)`. But at that point I think you might as well just use `Set<WeakRef<T>>` again.